dbms/src/Storages/DeltaMerge/ReadThread/SegmentReadTaskScheduler.cpp

```diff
@@ -213,57 +203,95 @@ bool SegmentReadTaskScheduler::isStop() const
+    for (auto itr = read_pools.begin(); itr != read_pools.end(); /**/)
     {
-        count++;
-        Stopwatch sw_sche_once;
-        MergedTaskPtr merged_task;
-        std::tie(merged_task, run_sche) = scheduleMergedTask();
-        if (merged_task != nullptr)
+        auto & pool = itr->second;
+        // No other component or thread hold this `pool`, we can release it.
+        // TODO: `weak_ptr` may be more suitable.
+        if (pool.use_count() == 1)
         {
-            auto elapsed_ms = sw_sche_once.elapsedMilliseconds();
-            if (elapsed_ms >= 5)
-            {
-                LOG_DEBUG(
-                    log,
-                    "scheduleMergedTask merged_task=<{}> cost={}ms pool_count={}",
-                    merged_task->toString(),
-                    elapsed_ms,
-                    pool_count);
-            }
-            SegmentReaderPoolManager::instance().addTask(std::move(merged_task));
+            LOG_INFO(pool->getLogger(), "Erase pool_id={}", pool->pool_id);
+            ++erased_pool_count;
+            itr = read_pools.erase(itr);
+            continue;
         }
-        if (!run_sche)
+        ++itr;
+
+        if (!needSchedule(pool))
         {
-            break;
+            ++sched_null_count;
+            continue;
         }
+
+        auto merged_task = scheduleMergedTask(pool);
+        if (merged_task == nullptr)
+        {
+            ++sched_null_count;
+            continue;
+        }
+        ++sched_succ_count;
+        SegmentReaderPoolManager::instance().addTask(std::move(merged_task));
+    }
+    return std::make_tuple(erased_pool_count, sched_null_count, sched_succ_count);
+}
+
+bool SegmentReadTaskScheduler::schedule()
+{
+    Stopwatch sw_sched_total;
+    std::lock_guard lock(mtx);
+    Stopwatch sw_do_sched;
+
+    auto pool_count = read_pools.size();
+    UInt64 erased_pool_count = 0;
+    UInt64 sched_null_count = 0;
+    UInt64 sched_succ_count = 0;
+    UInt64 sched_round = 0;
+    bool can_sched_more_tasks = false;
+    do
+    {
+        ++sched_round;
+        auto [erase, null, succ] = scheduleOneRound();
+        erased_pool_count += erase;
+        sched_null_count += null;
+        sched_succ_count += succ;
+
+        can_sched_more_tasks = succ > 0 && !read_pools.empty();
+        // If no thread is waitting to add tasks and there are some tasks to be scheduled, run scheduling again.
+        // Avoid releasing and acquiring `mtx` repeatly.
+        // This is common when query concurrency is low, but individual queries are heavy.
+    } while (add_waittings.load(std::memory_order_relaxed) <= 0 && can_sched_more_tasks);
+
+    if (read_pools.empty())
+    {
+        GET_METRIC(tiflash_storage_read_thread_counter, type_sche_no_pool).Increment();
     }
-    auto sche_all_elapsed_ms = sw_sche_all.elapsedMilliseconds();
-    if (sche_all_elapsed_ms >= 100)
+
+    auto total_ms = sw_sched_total.elapsedMilliseconds();
+    if (total_ms >= 100)
```

Comment:
Should we also log it when `sw_do_sched. elapsedMilliseconds()` cost more than 100ms?